### DISCUSSION TITLE
C++ and c code auto complete?

### DISCUSSION THREAD

**Discussion by nitzanmr** [thebaselab/codeapp#542] [Q&A]
First. I love you. Been waiting for way to code localiy on my ipad for the past year or so. And you just gave us the best of the best. so thank you.
I am trying to get auto complete like on vs code on the app for c++ and c but with no luck so far. It works on js (not sure if out of the box or something i installed) maybe you have a solution? 
Thank you for everything. And good luck. If you need any help i would love to help and try.

**Comment by nitzanmr** [comment]
Maybe this can help? 
https://github.com/invokr/clang-autocomplete

**Reply by bummoblizard** [reply]
There isn't auto complete for C or C++ for now. I think it's possible to implement but there could be performance issues.

**Reply by nitzanmr** [reply]
I”ll tell you what i think there is a large comunity of people that uses c and c++ for school work and univirsity i know only that in israel there is a real need for a place to run work on the ipad and i think this will be a good way to let people use the app for.. i can also try and implement the use of the app in my uni.. 
what preformence issues do you think will happen? 
thanks for the response. 
have a great week

**Reply by bummoblizard** [reply]
> I”ll tell you what i think there is a large comunity of people that uses c and c++ for school work and univirsity i know only that in israel there is a real need for a place to run work on the ipad and i think this will be a good way to let people use the app for.. i can also try and implement the use of the app in my uni.. 
> 
> what preformence issues do you think will happen? 
> 
> thanks for the response. 
> 
> have a great week 

Autocomplete usually works by calling the compiler repeatedly. Actually I tried to port a Python auto complete library before, the device gets a bit overheat and performance / stability seem to be limited by the iOS architecture.

**Reply by nitzanmr** [reply]
On what device did you test it? Old or new? Is it the only way to make autocomplete? In other ides’ it is all the same? 
is it the same in every ipad? Becuase of the need of processing intensive work?

**Reply by bummoblizard** [reply]
Yeah, I would say they are all pretty CPU and RAM intensive. I tested it on an iPad 8. Although there's different libraries for autocompletion, they are generally designed for desktop environment and don't work on an iOS device out of the box. (primarily due to the lack of subprocess support)

**Reply by nitzanmr** [reply]
Maybe it will be an option for the new m1 powered ipads. Beacuse of the m1 and the 8 gb of ram..
I have an ipad air a14 i think it won’’t be different then the ipad 8.
We can try if you want.

**Reply by nitzanmr** [reply]
Maybe something less intensive like adding pages with common commands for it to learn some of the options like it displays know? 
(When you have a file in the library it can get the commands you typed already).

**Reply by bummoblizard** [reply]
It's worth to revisit I guess. The goal is to find / implement a lightweight auto completion library that works Language Server Protocol (VS Code) and doesn't require subprocess support.

A server side approach like repl.it might also be an option but it takes extra efforts to set up and maintain the infrastructure.

**Reply by nitzanmr** [reply]
Why is there no subprocess support in ios? I haven’t developed for ios. Or in this field. Is there some way i can help you in that?

**Reply by bummoblizard** [reply]
Because it's prohibited by iOS. Systems calls like `fork()` will raise permission error.

**Reply by nitzanmr** [reply]
Is there any process manging tools like threads? Or is it also prohibited?

**Reply by bummoblizard** [reply]
Threads are available.

**Reply by nitzanmr** [reply]
Hi. I found some liteweight options maybe they will be good and stable? 
There is cquery and coc-clangd that i tried and it installed fine but i need to change the settings.json file to include it? 
If you have the time to check if it can be used. 
Thank you in advance nitzan :)

**Reply by nitzanmr** [reply]
Thanks for all your work again. nitzan :)
